src/yoroi-wallets/cardano/transactionManager.ts

```diff
@@ -0,0 +1,69 @@
+import {Storage} from '../storage'
+import {BackendConfig, Transactions} from '../types'
+import {TransactionCache} from './shelley'
+
+export const makeTransactionManager = async (storage: Storage, backendConfig: BackendConfig) => {
+  const transactionCache = await TransactionCache.create(storage.join('txs/'))
+  const memosManager = await makeMemosManager(storage.join('memos/'))
+
+  return {
+    // transactionCache api
+    getTransactions() {
+      const memos = memosManager.getMemos()
+      return Object.entries(transactionCache.transactions).reduce(
+        (result, [txId, tx]) => ({
+          ...result,
+          [txId]: {
+            ...tx,
+            memo: memos[txId] ?? null,
+          },
+        }),
+        {} as Transactions,
+      )
+    },
+    getPerRewardAddressCertificates() {
+      return transactionCache.perRewardAddressCertificates
+    },
+    getPerAddressTxs() {
+      return transactionCache.perAddressTxs
+    },
+    getConfirmationCounts() {
+      return transactionCache.confirmationCounts
+    },
+    clear: () => transactionCache.clear(),
+    resetState: () => transactionCache.resetState(),
+    subscribe: (handler) => transactionCache.subscribe(handler),
+    doSync: async (addressesByChunks: Array<Array<string>>) =>
+      transactionCache.doSync(addressesByChunks, backendConfig),
+
+    // memo api
+    saveMemo: (txId: string, memo: string) => memosManager.saveMemo(txId, memo),
+    clearMemos: () => memosManager.clear(),
+  } as const
+}
+
+export type TransactionManager = Awaited<ReturnType<typeof makeTransactionManager>>
+
+export const makeMemosManager = async (storage: Storage) => {
+  const getMemos = () => storage.getAllKeys().then(storage.multiGet).then(Object.fromEntries) ?? {}
+  let memos = await getMemos()
+
+  const saveMemo = async (txId: string, memo: string): Promise<void> => {
+    await storage.setItem(txId, memo)
+    memos = await getMemos()
+  }
```

Comment:
switch to optimistic update when saving, that we don't have to wait for the storage, make it readonly to avoid mutations

```suggestion
export const makeMemosManager = async (storage: Storage) => {
  const getMemos = () => storage.getAllKeys().then(storage.multiGet).then(Object.fromEntries)
  let memos: Readonly<....> = await getMemos()

  const updateMemos = (txId, memo) => memos = {...memos, {[txId]: memo}}  

  const saveMemo = async (txId: string, memo: string): Promise<void> => {
    updateMemos(txId, memo) 
    storage.setItem(txId, memo)
  }
```

Reply:
premature performance optimization
there is no way to know if the memos have changed until the resolve of the function

Reply:
where do we need to know if the memos have changed?